### PR TITLE
Fixed SharedStash for D2R, fixes Larzuk quest items reading

### PR DESCRIPTION
**Pull request by ktunkiewicz**
(so far only the "reading" part and tests, didn't have time to work on the "writing" part yet.)

Credit goes to Silospen for this commit in GoMule: https://sourceforge.net/p/gomule/git/ci/f1a0b1ce069b5c475c4d1b45a9ab10c92e732755

This fixes:
- Shared stash in D2R now reads properly
- Chars with customized items now read properly

### CODE REVIEW

**Comment by ktunkiewicz** [comment]
I'll write the "writing" part some time soon, just wanted to PR this early in case you have any comments

**Comment by dschu012** [comment]
Looks good to me. Add I'll wait for writing to merge it.

File: tests/d2/d2s.spec.ts [hunk 1]
```diff
@@ -1,147 +1,163 @@
...
+  it("should read version 98 complex character 3", async () => {
+    const inputstream = fs.readFileSync(path.join(__dirname, "../../examples/chars/98/PaladinTwoNormal.d2s"));
+    const save = await read(inputstream, constants);
+    //console.log(JSON.stringify(save, null, 2));
+    expect(save.header.name).to.eq("PaladinTwo");
+    expect(save.attributes.strength).to.eq(159);
+    expect(save.items.length).to.eq(73);
+  });
+
+  it("should write version 98 complex character", async () => {
```
Comment (ktunkiewicz) [inline tests/d2/d2s.spec.ts:40]: not sure why git re-wrote the whole file, but this is the only new test

**Comment by ktunkiewicz** [comment]
Still shared stash writing needs to be finished, I'll be working on that soon.

**Comment by ktunkiewicz** [comment]
I think it is done now. Added test that read, then parse, write and re-read - to test if the data that was written is readable and correct.

Also added `hardcore: boolean` to the stash type, so it can be detected if that is softcore or hardcore stash.

**Review by dschu012 — APPROVED** [review]

**Pull request merged**